src/utils/path.c

```diff
@@ -0,0 +1,389 @@
+#if defined(_WIN32) || defined(WIN32)
+    path tmp = convert_to_windows_slashes(p);
+    return PathFileExistsA(tmp.buf) == TRUE;
+#else
+    return access(p->buf, F_OK) == 0;
+#endif
+}
+
+static bool find_ext_start_end(const str *tmp, size_t *start, size_t *end) {
+    size_t spos, dpos;
+    if(str_last_of(tmp, '/', &spos)) {
+        spos += 1;
+    } else {
+        spos = 0;
+    }
+    if(!str_last_of(tmp, '.', &dpos)) {
+        return false;
+    }
+    if(spos >= dpos) {
+        return false;
+    }
+    *start = dpos;
+    *end = str_size(tmp);
+    return true;
+}
+
+void path_ext(const path *path, str *dst) {
+    str tmp;
+    size_t start, end;
+    str_from_c(&tmp, path->buf);
+    if(find_ext_start_end(&tmp, &start, &end)) {
+        str_from_slice(dst, &tmp, start, end);
+    } else {
+        str_create(dst);
+    }
+    str_free(&tmp);
+}
+
+void path_set_ext(path *path, const char *ext) {
+    str tmp;
+    size_t start, end;
+    str_from_c(&tmp, path->buf);
+    if(find_ext_start_end(&tmp, &start, &end)) {
+        str_cut(&tmp, str_size(&tmp) - start);
+    }
+    str_append_c(&tmp, ext);
+    path_from_str(path, &tmp);
+    ENSURE_ZERO(path->buf);
+    str_free(&tmp);
+}
+
+void path_dossify_filename(path *path) {
+    str tmp;
+    size_t pos = 0;
+    str_from_c(&tmp, path->buf);
+    str_last_of(&tmp, '/', &pos);
+    for(size_t i = pos + 1; i < str_size(&tmp); i++) {
+        const char c = str_at(&tmp, i);
+        if(c == '.') {
+            continue;
+        }
+        str_set_at(&tmp, i, isalnum(c) ? toupper(c) : '_');
+    }
+    path_from_str(path, &tmp);
+    ENSURE_ZERO(path->buf);
+    str_free(&tmp);
+}
+
+void path_stem(const path *path, str *dst) {
+    str tmp;
+    size_t start, end;
+    str_from_c(&tmp, path->buf);
+    if(str_last_of(&tmp, '/', &start)) {
+        start += 1;
+    } else {
+        start = 0;
+    }
+    if(!str_last_of(&tmp, '.', &end) || end <= start) {
+        end = str_size(&tmp);
+    }
+    if(start < end) {
+        str_from_slice(dst, &tmp, start, end);
+    } else {
+        str_create(dst);
+    }
+    str_free(&tmp);
+}
+
+void path_parents(const path *path, str *dst) {
+    str tmp;
+    size_t end;
+    str_from_c(&tmp, path->buf);
+    if(str_last_of(&tmp, '/', &end) && end > 1) {
+        str_from_slice(dst, &tmp, 0, end);
+    } else {
+        str_create(dst);
+    }
+    str_free(&tmp);
+}
+
+void path_filename(const path *path, str *dst) {
+    str tmp;
+    size_t start;
+    str_from_c(&tmp, path->buf);
+    const size_t end = str_size(&tmp);
+    if(str_last_of(&tmp, '/', &start)) {
+        start += 1;
+    } else {
+        start = 0;
+    }
+    if(start < end) {
+        str_from_slice(dst, &tmp, start, end);
+    } else {
+        str_create(dst);
+    }
+    str_free(&tmp);
+}
+
+bool path_touch(const path *path) {
+    FILE *fp = fopen(path->buf, "w");
+    if(fp == NULL) {
+        return false;
+    }
+    fclose(fp);
+    return true;
+}
```

Comment:
This `path_touch` is very unlike touch(1):
- It empties the file
- It cannot operate on directories.

path_touch is currently unused, and barely tested.
Perhaps it should be removed or renamed?

Reply:
It's used in tests, which is why it was added here... It might be better to just move it as a utility function in the test file.